File: c++/SJF.cpp
#include <iostream>
using namespace std;

class SJF {
   public:
    int n, *BT, *WT, *TAT, *PID;
    float awt = 0, atat = 0;
    SJF(int n) : n(n) {
        BT = new int[n];
        WT = new int[n];
        TAT = new int[n];
        PID = new int[n];
        for (int i = 0; i < n; i++) {
            cout << "Enter burst time of process - " << i + 1
                 << " : ";
            cin >> BT[i];
            PID[i] = i + 1;
        }
    }
    void swap(int &x, int &y) {
        int temp = x;
        x = y;
        y = temp;
    }
    void sort() {
        for (int i = 0; i < n - 1; i++) {
            int min = i;
            for (int j = i + 1; j < n; j++) {
                if (BT[min] > BT[j]) min = j;
            }
            swap(BT[i], BT[min]);
            swap(PID[i], PID[min]);
        }
    }
    void calculate() {
        WT[0] = 0;
        TAT[0] = BT[0];
        atat = TAT[0];
        for (int i = 1; i < n; i++) {
            WT[i] = WT[i - 1] + BT[i - 1];
            TAT[i] = BT[i] + WT[i];
            atat += TAT[i];
            awt += WT[i];
        }
    }
    void print_details() {
        cout << "\nProcess\tBT\tWT\tTAT\n";
        for (int i = 0; i < n; i++) {
            printf("%d\t%d\t%d\t%d\n", PID[i], BT[i], WT[i], TAT[i]);
        }
        cout << "Avg waiting time and TAT are: " << awt / n << " "
             << atat / n;
    }
    ~SJF() {
        delete[] BT;
        delete[] WT;
        delete[] TAT;
        delete[] PID;
        cout << "\nDeleted\n";
    }
};

int main() {
    int n;
    cout << "Enter number of processes: ";
    cin >> n;
    SJF s(n);
    s.sort();
    s.calculate();
    s.print_details();
    return 0;
}

File: java/SJF.java
package java;
import java.util.Scanner;

class SJF {
    int n, burstTime[], processID[], waitingTime[], turnAroundTIme[];
    float averageTurnAroundTime, averageWaitingTime;

    SJF(Scanner scan) {
        System.out.print("Enter number of processes: ");
        n = scan.nextInt();
        burstTime = new int[n];
        waitingTime = new int[n];
        turnAroundTIme = new int[n];
        processID = new int[n];
        for (int i = 0; i < n; i++) {
            System.out.print("Enter burst time of process - " + (i + 1) + " : ");
            burstTime[i] = scan.nextInt();
            processID[i] = i + 1;
        }
    }

    void sort() {
        for (int i = 0; i < n - 1; i++) {
            int min = i;
            for (int j = i + 1; j < n; j++) {
                if (burstTime[min] > burstTime[j]) {
                    min = j;
                }
            }
            int temp = burstTime[min];
            burstTime[min] = burstTime[i];
            burstTime[i] = temp;
            temp = processID[min];
            processID[min] = processID[i];
            processID[i] = temp;
        }
    }

    void calculate() {
        waitingTime[0] = 0;
        turnAroundTIme[0] = burstTime[0];
        averageTurnAroundTime = turnAroundTIme[0];
        for (int i = 1; i < n; i++) {
            waitingTime[i] = waitingTime[i - 1] + burstTime[i - 1];
            turnAroundTIme[i] = burstTime[i] + waitingTime[i];
            averageWaitingTime += waitingTime[i];
            averageTurnAroundTime += turnAroundTIme[i];
        }
    }

    void printDetails() {
        System.out.println("PROCSESS burstTime waitingTime turnAroundTIme");
        for (int i = 0; i < n; i++) {
            System.out.println(String.format("%d\t %d  %d  %d", processID[i], burstTime[i], waitingTime[i], turnAroundTIme[i]));
        }
        System.out.println("Avg watiing time & turn around time are " + (averageWaitingTime / n) + " " + (averageTurnAroundTime / n));
    }

    public static void main(String args[]) {
        Scanner scan = new Scanner(System.in);
        SJF w = new SJF(scan);
        w.sort();
        w.calculate();
        w.printDetails();
    }
}